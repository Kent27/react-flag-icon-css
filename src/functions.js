// @flow
import type classes from 'classnames'
import type { KeyCallbackType, CountryType } from './types/flow'
import oCountries from './countries'


export const getCountries = () : Array<CountryType> => oCountries
export const getCountryCodes = (): Array<string> =>
  getCountries().map((o: CountryType) : string => o.code)

export const objectKeysApplyFn = (obj: classes, // eslint-disable-line import/prefer-default-export, max-len
                                  fn: KeyCallbackType) : classes =>
  Object.keys(obj).reduce((retObj: classes, key: string) : classes => { // eslint-disable-line arrow-body-style, max-len
    return {
      ...retObj,
      [fn(key)]: obj[key]
    }
  }, {})
