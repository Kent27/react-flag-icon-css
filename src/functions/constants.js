export const baseThemeStyleName = 'theme-base'
export const flagIconClassesPrefixName = 'flag-icon'
export const flagIconClassesPrefix = `${flagIconClassesPrefixName}-`
