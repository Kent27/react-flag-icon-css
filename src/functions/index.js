// @flow
import type classes from 'classnames'
import type { KeyCallbackType } from '../types/flow'

export { default as makeClassnames, makeClassesObject } from './classnames'
export { default as makeFlagIconOptions } from './options'
export { default as makeStyles } from './styles'
export * as countries from './countries'
export * as constants from './constants'


export const objectKeysApplyFn = (obj: classes, fn: KeyCallbackType): classes =>
  Object.keys(obj).reduce(
    (rObj: classes, key: string): classes => ({ ...rObj, [fn(key)]: obj[key] })
    , {}, // reduce() second argument: initial value of rObj
  )

export const diffArrays = <T>(arA: Array<T>, arB: Array<T>): Array<T> => {
  const arrays = (arA.length > arB.length) ? [arA, arB] : [arB, arA]
  return arrays[0].filter((newArElem: T): boolean => !arrays[1].includes(newArElem))
}
