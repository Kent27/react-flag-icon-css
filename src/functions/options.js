import type { FlagIconOptionsType } from '../types/flow'
import { FlagIconOptionsTypeTcomb } from '../types/tcomb'


const defaultOptions: FlagIconOptionsType = {
  useCssModules: true
}

export default (options?: FlagIconOptionsType) : FlagIconOptionsType =>
  FlagIconOptionsTypeTcomb({ ...defaultOptions, ...options })
