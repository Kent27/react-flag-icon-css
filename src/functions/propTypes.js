// @flow
type ObjStringKeyMixedValueType = { [string]: mixed }
const validatorKeyName = '__exact__'

/**
 * This 'custom validator' function is called by the `prop-types` module at runtime.
 * If `propsValues`, the Object containing the 'props' passed to a React component
 * or to `PropTypes.checkPropTypes()`, contains a 'prop' (= one of its keys) that
 * is not in `propsObject` (the 'props' Object assigned by the user to
 * SomeComponent.propTypes or passed to `PropTypes.checkPropTypes()` - to declare
 * the prop-types), `prop-types` prints the text of `Error` to the console.
 *
 * SEE: https://github.com/reactjs/prop-types
 */
const fnExactValidator = (
  // prop-types Object supplied by the user. i.e {lorem: PropTypes.string}
  propsObject: ObjStringKeyMixedValueType,
  // prop-values Object supplied by the user. i.e {lorem: 'ipsum'}
  propsValues: ObjStringKeyMixedValueType,
  propName: string,
  componentName: string,
  // eslint-disable-next-line consistent-return
): ?Error => {
  const passedPropNames: string[] = Object.keys(propsValues)
  const extra = passedPropNames.filter(
    (prop: string) => !Object.prototype.hasOwnProperty.call(propsObject, prop),
  )

  if (extra.length > 0) {
    return new Error(
      `Invalid prop(s) ${JSON.stringify(extra)} supplied to \
\`${componentName}\`, expected one of [${passedPropNames.join(',')}].`,
    )
  }
}

/**
 * Adds a 'custom validator' function to `obj`; this function makes sure that any
 * 'prop' passed by the user at runtime exists in `obj`, otherwise it throws an Error.
 * `obj` is a 'props' object, the one you would assign to `SomeComponent.propTypes`
 * or pass to `PropTypes.checkPropTypes()`. i.e {lorem: PropTypes.string}
 *
 * We use bind() to make a new `fnExactValidator` function with the first argument
 * set as: `obj`. We need to do this because `prop-types` doesn't pass this object
 * to 'custom validator' functions such as `fnExactValidator`.
 *
 * This type of validation is the runtime equivalent to Flow's 'exact object types'.
 * SEE: https://flow.org/en/docs/types/objects/#toc-exact-object-types
 */
// eslint-disable-next-line import/prefer-default-export, max-len
export const AddExactValidator = <T: ObjStringKeyMixedValueType>(obj: T): T => {
  if (Object.prototype.hasOwnProperty.call(obj, validatorKeyName)) {
    // TODO: print a message or throw when not in production?
    return obj // The custom validator is already 'installed'.
  }

  return {
    ...obj,
    [validatorKeyName]: fnExactValidator.bind(this, obj),
  }
}
