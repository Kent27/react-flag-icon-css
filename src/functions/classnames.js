// @flow
import classnames from 'classnames'
import type classes from 'classnames'
import type { FlagIconPropsType, FlagIconOptionsType } from '../types/flow'
import { baseThemeStyleName, flagIconClassesPrefix, flagIconClassesPrefixName } from './constants'


export const makeClassesObject = ({ ...props,
                                       code,
                                       flip = '',
                                       size = '',
                                       squared = false,
                                       rotate }: FlagIconPropsType,
                                       options: FlagIconOptionsType) : classes => {
  let obj = {
    [flagIconClassesPrefixName]: true,
    [`${flagIconClassesPrefix}squared`]: squared,
    [`${flagIconClassesPrefix}${size}`]: size,
    [`${flagIconClassesPrefix}${code}`]: code,
    [`${flagIconClassesPrefix}flip-${flip}`]: flip,
    [`${flagIconClassesPrefix}rotate-${rotate ? rotate : ''}`]: rotate, // eslint-disable-line no-unneeded-ternary
    [baseThemeStyleName]: options.useCssModules && options.themeStyles
  }

  if (props.className && options.useCssModules) {
    obj = { ...obj, [props.className]: true }
  }
  return obj
}

export default (props: FlagIconPropsType, options: FlagIconOptionsType) : string =>
  classnames(makeClassesObject(props, options))
