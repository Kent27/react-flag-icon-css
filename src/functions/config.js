import fs from 'fs'
import path from 'path'
import type { FlagIconCodeType } from '../types/flow'
import { flagIconModulePath, flagIconClassesPrefix } from '../functions/constants'
import styles from '../styles'


export const GetFlagIconModuleCountryCodes = ( // eslint-disable-line import/prefer-default-export
    modulePath: string = flagIconModulePath
): boolean =>
  fs.readdirSync(path.resolve(`${modulePath}/flags/4x3`)).reduce(
    (retAr: [], fileName: string): FlagIconCodeType[] => { // reduce() first argument: function
      const code = path.basename(fileName, path.extname(fileName))
      return styles[`${flagIconClassesPrefix}${code}`] ? [...retAr, code] : retAr
    }
    , [] // reduce() second argument: initial value of retAr
  )
