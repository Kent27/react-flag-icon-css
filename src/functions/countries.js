// @flow
import { ThrowNoJsonLoaderError } from './config'
import type { CountryType, FlagIconCodeType } from '../types/flow'
import oCountries from '../../static/countries.json'

// Webpack 2 can require/import json out of the box, Webpack 1 and maybe other
// bundlers cannot, so we warn the user. UglifyJS strips this code out in production.
if (process.env.NODE_ENV !== 'production') {
  if (!oCountries || Object.keys(oCountries).length === 0) {
    ThrowNoJsonLoaderError()
  }
}

export const getCountries = (): CountryType[] => oCountries
export const getCountryCodes = (): FlagIconCodeType[] =>
  getCountries().map((o: CountryType): FlagIconCodeType => o.code)
