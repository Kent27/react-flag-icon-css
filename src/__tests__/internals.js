// @flow
// Test functions used in tests.
import React from 'react'
import test from 'ava' // eslint-disable-line import/no-extraneous-dependencies
import render from 'react-test-renderer' // eslint-disable-line import/no-extraneous-dependencies
import { restoreClassNamesInTree, makeFlagIcons } from './internals/functions'
import { cssModulePrefix } from './internals/constants'

test('restoreClassNamesInTree', t => {
  // Sample 'react-test-renderer' render(<SomeComponent />) output.
  const tree = {
    type: 'div',
    props: { className: `${cssModulePrefix}some-class`, lorem: 'ipsum' },
    children: {
      type: 'span',
      props: {
        className: `${cssModulePrefix}some-other-class ${cssModulePrefix}ipsum`,
      },
    },
  }

  // We expect `demodulifiedTree` to be `tree` with the `cssModulePrefix` string
  // removed from every `className`.
  const demodulifiedTree = restoreClassNamesInTree(tree)
  t.snapshot(demodulifiedTree)
})

test('makeFlagIcons', t => {
  const options = [{ useCssModules: false }, { useCssModules: true }]
  const props = [{ code: 'it', flip: 'horizontal' }, { code: 'es', rotate: 60 }]

  // $FlowExpectError
  t.throws(() => makeFlagIcons())
  // $FlowExpectError
  t.throws(() => makeFlagIcons(options))
  t.throws(() => makeFlagIcons(options, [{ code: 'it' }]))

  const [FlagIcon, FlagIconModules] = makeFlagIcons(options, props)
  const App = (
    <div>
      {FlagIcon}
      {FlagIconModules}
    </div>
  )
  const tree = render.create(App).toJSON()
  t.snapshot(tree)
})
