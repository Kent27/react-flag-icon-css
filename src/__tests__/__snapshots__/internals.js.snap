// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeFlagIcons 1`] = `
<div>
  <span
    className="flag-icon flag-icon-it flag-icon-flip-horizontal"
  />
  <span
    className="__MODULE__flag-icon __MODULE__flag-icon-es __MODULE__flag-icon-rotate-60"
  />
</div>
`;

exports[`restoreClassNamesInTree 1`] = `
Object {
  "children": Object {
    "props": Object {
      "className": "some-other-class ipsum",
    },
    "type": "span",
  },
  "props": Object {
    "className": "some-class",
    "lorem": "ipsum",
  },
  "type": "div",
}
`;
